README.md
# WK2124 软件包

## 1 介绍

WK2124 软件包是为[WK2124 SPI转四串口扩展芯片](http://www.wkmic.com/News_Show.php?theId=17)而开发的RT-Thread驱动包。通过`RT-Thread SPI` 设备和`RT-Thread UART` 设备驱动框架，为RT-Thread系统新增最多四路串口字符设备。与片上串口使用方式一致，实现了对串口设备进行open、close、read、write、control操作的功能，并支持轮询和中断接收两种模式。

WK2124 软件包为串口设备资源不足的产品提供串口扩展的方法，目前已在stm32f429验证通过。


### 1.1 目录结构

```shell
wk2124-latest/
├── ChangeLog.md				# 修改记录
├── LICENSE						# 软件包许可证
├── README.md					# 软件包使用说明
├── sample						# 串口使用程序样例
│   ├── uart_int_sample.c		# 中断模式
│   └── uart_sample.c			# 轮询模式
├── SConscript					# RT-Thread 默认的构建脚本
├── wk2124s.c					# wk2124 驱动
├── wk2124s.h					# wk2124 驱动头文件
├── wk2124_usart.c				# wk2124 spi及uart驱动文件
├── wk2124_usart_v2.c			# wk2124 spi及uart_v2驱动文件
└── wk2124_usart.h				# wk2124 spi及uart驱动头文件
```

### 1.2 许可证

AT24CXX 软件包遵循  Apache-2.0 许可，详见 LICENSE 文件。

### 1.3 依赖

- RT-Thread 4.0.1+
- SPI 驱动：WK2124 设备使用 SPI 进行数据通讯，需要系统 SPI 驱动框架支持；
- PIN 驱动：用于处理设备中断引脚；



## 2 获取软件包

使用 `wk2124` 软件包需要在 RT-Thread 的包管理器中选择它，具体路径如下：

```
RT-Thread online packages
    peripheral libraries and drivers  --->
        [*] wk2124: spi wk2124 driver library.  --->
           --- wk2124: spi wk2124 driver library.
            [*]   Enable UART SWK1
            [*]   Enable UART SWK2
            [*]   Enable UART SWK3
            [*]   Enable UART SWK4
            WK2124 device configure  --->
            	(spi2dev) SPI device name
            	(17) IRQ pin number
            	(11059200) Set crystal frequency. Unit Hz
            Version (latest)  --->
```


每个功能的配置说明如下：

- Enable UART SWK1：使能串行字符设备uart swk1
- Enable UART SWK2：使能串行字符设备uart swk2
- Enable UART SWK3：使能串行字符设备uart swk3
- Enable UART SWK4：使能串行字符设备uart swk4
- WK2124 device configure：配置使用设备的参数
  - SPI device name：配置使用 SPI 的设备名称（注意需设置为**非 SPI 总线设备**）
  - IRQ pin number：配置设备连接的中断引脚号（根据实际使用引脚号修改）
  - Set crystal frequency：wk2124硬件电路的晶振频率（影响串口波特率）
- Version：配置软件包版本，默认最新版本。

然后让 RT-Thread 的包管理器自动更新，或者使用 `pkgs --update` 命令更新包到 BSP 中。



## 3 使用 wk2124 软件包

按照前文介绍，获取 wk2124 软件包并配置参数项。软件包编译进RT-Thread系统以后，通过宏定义INIT_ENV_EXPORT(wk2124_device_init)，软件包会在开机时自动初始化，并产生串行字符设备。

```shell
# 初始化信息
\ | /
- RT -     Thread Operating System
 / | \     4.0.1 build Aug 25 2020
 2006 - 2019 Copyright by rt-thread team
.....
[I/drv] RT-Thread WK2124 package initialize success.
......
msh />

# 串行字符设备
msh />list_device
device           type         ref count
-------- -------------------- ----------
......
uartswk4 Character Device     0       
uartswk3 Character Device     0       
uartswk2 Character Device     0       
uartswk1 Character Device     0
......
```

uartswk1--4串行字符设备支持轮询和中断接收两种模式（不支持DMA模式），使用方法可参考[RT-Thread UART设备使用文档](https://www.rt-thread.org/document/site/programming-manual/device/uart/uart/)和[sample](sample/)文件夹。



## 4 注意事项

- 获取软件包时，需要注意正确配置使用的 SPI 设备名称、中断引脚号和芯片晶振频率；
- 推荐使用最新版，V1.0.0波特率计算存在BUG，且wk2124的串口关闭不彻底。
- _v2版本为适配RT-Thread serial_v2版本。


## 5 联系方式

* 维护：[liyaohong(MrMichael)](https://github.com/MrMichael)
* 主页：https://github.com/MrMichael/wk2124.git
* 邮箱：1035285359@qq.com



ChangeLog.md
2020-08-26:

1. Release version 1.0.0

2022-04-07：

1. Release version 2.0.0

wk2124s.c
/*
 *WK2124S 支持 SPI模式
 *WK2124S IRQ中断信号的输出低电平有效
 */

#include "wk2124s.h"

#ifdef PKG_USING_WK2124

void EXHW_WK2124_Write_Reg(struct rt_spi_device *device, uint8_t reg, uint8_t dat)
{
    static uint8_t sendbuf[2];

    sendbuf[0] = reg;
    sendbuf[1] = dat;

    rt_spi_send(device, sendbuf, 2);
}

uint8_t EXHW_WK2124_Read_Reg(struct rt_spi_device *device, uint8_t reg)
{
    static uint8_t sendbuf[2];
    static uint8_t recbuf[2];
    
    sendbuf[0] = 0x40 | reg;
    sendbuf[1] = 0x00;
    rt_spi_transfer(device, sendbuf, recbuf, 2);
    return recbuf[1];
}

void EXHW_WK2124_Write_FIFO(struct rt_spi_device *device, uint8_t reg, uint8_t *buf, uint16_t len)
{
    uint16_t cnt = 0;
    uint8_t sendbuf[257];

    sendbuf[0] = 0x80 | reg;
    for(cnt = 0; (cnt < len)&&(cnt < 256); cnt++) {
        sendbuf[1+cnt] = buf[cnt];
    } 
    rt_spi_send(device, sendbuf, cnt+1);
}

void EXHW_WK2124_Read_FIFO(struct rt_spi_device *device, uint8_t reg, uint8_t *buf, uint16_t len)
{
    uint16_t cnt = 0;
    uint8_t sendbuf[257];
    uint8_t recbuf[257];
    uint16_t i = 0;

    sendbuf[0] = 0xC0 | reg;
    for(cnt = 0; (cnt < len)&&(cnt < 256); cnt++) {
        sendbuf[1+cnt] = 0x00;
    } 
    rt_spi_transfer(device, sendbuf, recbuf, cnt+1);
    for ( i = 0; i < cnt; i++) {
        buf[i] = recbuf[1+i];
    }
}

uint16_t Wk2124_SendBuf(struct rt_spi_device *device, uint8_t index, uint8_t *sendbuf,uint16_t len)
{
    uint16_t ret = 0,tfcnt = 0,sendlen = 0;
    uint8_t  fsr = 0;
    
    fsr = EXHW_WK2124_Read_Reg(device, SPAGE0_FSR(index));
    if(~fsr & 0x02 )//子串口发送FIFO未满
    {
        tfcnt = EXHW_WK2124_Read_Reg(device, SPAGE0_TFCNT(index));//读子串口发送fifo中数据个数
        sendlen = 256 - tfcnt;//FIFO能写入的最多字节数

        if(sendlen < len){
            ret = sendlen; 
            EXHW_WK2124_Write_FIFO(device, SPAGE0_FDAT(index),sendbuf,sendlen);
        }else{
            EXHW_WK2124_Write_FIFO(device, SPAGE0_FDAT(index),sendbuf,len);
            ret = len;
        }
    }
    return ret;
}

uint16_t Wk2124_GetBuf(struct rt_spi_device *device, uint8_t index, uint8_t *getbuf, uint16_t len)
{
    uint16_t ret=0, rfcnt = 0;
    uint8_t fsr = 0;
    
    fsr = EXHW_WK2124_Read_Reg(device, SPAGE0_FSR(index));
    if(fsr & 0x08 )//子串口接收FIFO未空
    {
        rfcnt = EXHW_WK2124_Read_Reg(device, SPAGE0_RFCNT(index));//读子串口发送fifo中数据个数
        if(rfcnt == 0)//当RFCNT寄存器为0的时候，有两种情况，可能是256或者是0，这个时候通过FSR来判断，如果FSR显示接收FIFO不为空，就为256个字节
        {
            rfcnt = 256;
        }
        if (rfcnt < len)
        {
            EXHW_WK2124_Read_FIFO(device, SPAGE0_FDAT(index), getbuf, rfcnt);
            ret = rfcnt;
        } else {
            EXHW_WK2124_Read_FIFO(device, SPAGE0_FDAT(index), getbuf, len);
            ret = len;
        }

    }
     return ret;    
}

void EXHW_WK2124_Init(struct rt_spi_device *device)
{
    /*使能子串口1,2,3,4的时钟*/
    EXHW_WK2124_Write_Reg(device, GENA,0x0F);
    
    /*复位子串口1,2,3,4*/
    EXHW_WK2124_Write_Reg(device, GRST,0x0F);
    
    /*使能子串口1,2,3,4的全局中断 */
    EXHW_WK2124_Write_Reg(device, GIER,0x0F);
    
}

void EXHW_WK2124_Disable_Tx(struct rt_spi_device *device, uint8_t index)
{
    uint8_t scr = 0;

    scr = EXHW_WK2124_Read_Reg(device, SPAGE0_SCR(index)); 
    scr &= ~(1 << 1);
    EXHW_WK2124_Write_Reg(device, SPAGE0_SCR(index),scr);
}

void EXHW_WK2124_Enable_Tx(struct rt_spi_device *device, uint8_t index)
{
    uint8_t scr = 0;

    scr = EXHW_WK2124_Read_Reg(device,SPAGE0_SCR(index)); 
    scr |= (1 << 1);
    EXHW_WK2124_Write_Reg(device,SPAGE0_SCR(index),scr);
}

void EXHW_WK2124_Disable_Rx(struct rt_spi_device *device, uint8_t index)
{
    uint8_t scr = 0;
    scr = EXHW_WK2124_Read_Reg(device,SPAGE0_SCR(index)); 
    scr &= ~(1 << 0);
    EXHW_WK2124_Write_Reg(device,SPAGE0_SCR(index),scr);
}

void EXHW_WK2124_Enable_Rx(struct rt_spi_device *device, uint8_t index)
{
    uint8_t scr = 0,fcr = 0;
    
    //复位 n 串口的FIFO
    fcr = EXHW_WK2124_Read_Reg(device,SPAGE0_FCR(index)); 
    fcr |= (1 << 0);
    EXHW_WK2124_Write_Reg(device,SPAGE0_FCR(index),fcr);
    //使能串口 n  的接收
    scr = EXHW_WK2124_Read_Reg(device,SPAGE0_SCR(index)); 
    scr |= (1 << 0);
    EXHW_WK2124_Write_Reg(device,SPAGE0_SCR(index),scr);
}

#endif


sample/uart_sample.c
#include <rtthread.h>

#define SAMPLE_UART_NAME       "uartswk2"      /* 串口设备名称 */

/* 用于接收消息的信号量 */
static struct rt_semaphore rx_sem;
static rt_device_t serial;

/* 接收数据回调函数 */
static rt_err_t uart_input(rt_device_t dev, rt_size_t size)
{
    /* 串口接收到数据后产生中断，调用此回调函数，然后发送接收信号量 */
    rt_sem_release(&rx_sem);
    return RT_EOK;
}


static int uart_sample(int argc, char *argv[])
{
    rt_err_t ret = RT_EOK;
    char uart_name[RT_NAME_MAX];
    char str[] = "hello RT-Thread!\r\n";
    char ch;

    if (argc == 2)
    {
        rt_strncpy(uart_name, argv[1], RT_NAME_MAX);
    }
    else
    {
        rt_strncpy(uart_name, SAMPLE_UART_NAME, RT_NAME_MAX);
    }

    /* 查找串口设备 */
    serial = rt_device_find(uart_name);
    if (!serial)
    {
        rt_kprintf("find %s failed!\n", uart_name);
        return RT_ERROR;
    }

    /* 以读写及中断接收方式打开串口设备 */
    rt_device_open(serial, RT_DEVICE_OFLAG_RDWR);
    /* 发送字符串 */
    rt_device_write(serial, 0, str, (sizeof(str) - 1));

    while (1) {
        rt_thread_mdelay(100);
        if (rt_device_read(serial, -1, &ch, 1)) {
            rt_kprintf("%s: %c\n", uart_name, ch);
            /* 发送已接收的数据 */
            rt_device_write(serial, 0, &ch, 1);
        }
    }

    return ret;
}

/* 导出到 msh 命令列表中 */
MSH_CMD_EXPORT(uart_sample, uart device sample);


wk2124_usart.c
#include "board.h"
#include <rtdevice.h>
#include <rthw.h>
#include <rtthread.h>

#include "wk2124_usart.h"
#include "wk2124s.h"

#include <drv_log.h>
// #define DRV_DEBUG


#ifdef PKG_USING_WK2124

//WK2124晶振频率，单位Hz
#ifndef WK2124_Fosc
    #define WK2124_Fosc     11059200     
#endif

#ifndef WK2124_IRQ_PIN
    #define WK2124_IRQ_PIN        17  /* PB1 */
#endif

struct rt_spi_device *wk2124_device = RT_NULL;

/* wk2124 uart driver */
struct wk2124_uart 
{
    uint8_t swk_index; 
    uint8_t irq_enable;
    struct rt_spi_device *spi_device;
};

/* 用于接收中断的信号量 */
static struct rt_semaphore irq_sem;

static rt_err_t wk2124_configure(struct rt_serial_device *serial,
                              struct serial_configure *cfg)
{
    struct wk2124_uart *uart;
    uint16_t baudrate;
    uint8_t baudrate_h, baudrate_l, baudrate_dec, spage0_lcr;

    RT_ASSERT(serial != RT_NULL);
    RT_ASSERT(cfg != RT_NULL);
    uart = (struct wk2124_uart *)serial->parent.user_data;

    /*切换到PAGE0页中的子串口寄存器组 */
    EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE(uart->swk_index),0x00);

    /*子串口 1 控制寄存器 */
    EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE0_SCR(uart->swk_index),0x00); //子串口 发送使能  接收使能

    spage0_lcr = 0;
    if (cfg->stop_bits > 0) {
        spage0_lcr |= 0x01;
    }
    if (cfg->parity == 1) {
        spage0_lcr &= 0xfb;
        spage0_lcr |= 0x02;
    } else if (cfg->parity == 2) {
        spage0_lcr &= 0xfd;
        spage0_lcr |= 0x04;
    }
    if (cfg->data_bits == 9) {
        spage0_lcr |= 0x08;
    }
    /*子串口 配置寄存器*/
    EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE0_LCR(uart->swk_index),spage0_lcr);    //子串口 正常输出,普通模式,8位数据位,0校验,1位停止位

    /*子串口 FIFO控制寄存器*/
    EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE0_FCR(uart->swk_index),0x03);    //子串口 发送触发点,接收触发点 
                                                                                  //使能 发送,接收FIFO 复位发送接收FIFO
    /*子串口 中断使能寄存器*/
    EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE0_SIER(uart->swk_index),0x00); //子串口 禁止接收FIFO数据错误中断
                                                                                                //禁止发送FIFO空中断
                                                                                                //禁止发送FIFO触点中断
                                                                                                //禁止接收FIFO接收超时中断
                                                                                                //禁止接收FIFO接收触点中断

    baudrate = WK2124_Fosc/cfg->baud_rate/16;
    baudrate_h = baudrate/0x100;
    baudrate_l = baudrate%0x100 -1;
    baudrate_dec = (uint8_t)(((WK2124_Fosc/cfg->baud_rate/16.0f) - baudrate)*16);

    /*切换到PAGE1页中的子串口寄存器组 */
    EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE(uart->swk_index),0x01);

    /*子串口1 波特率配置寄存器高字节 [Reg = 11.0592/(115200*16) = 6] */
    EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE1_BAUD1(uart->swk_index),baudrate_h);
    /*子串口1 波特率配置寄存器低字节 */
    EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE1_BAUD0(uart->swk_index),baudrate_l);
    /*子串口1 波特率配置寄存器小数部分*/
    EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE1_PRES(uart->swk_index),baudrate_dec);

    /*切换到PAGE0页中的子串口寄存器组 */
    EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE(uart->swk_index),0x00);

    return RT_EOK;
}

static rt_err_t wk2124_control(struct rt_serial_device *serial,
                            int cmd, void *arg)
{
    struct wk2124_uart *uart;
    RT_ASSERT(serial != RT_NULL);
    uart = (struct wk2124_uart *)serial->parent.user_data;
    switch (cmd)
    {
    case RT_DEVICE_CTRL_CLR_INT:
        // /* disable rx irq */
        /*切换到PAGE0页中的子串口寄存器组 */
        EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE(uart->swk_index),0x00);
        /*子串口 控制寄存器 */
        EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE0_SCR(uart->swk_index),0x00);	//子串口 发送不使能  接收不使能
        /*子串口 FIFO控制寄存器*/
        EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE0_FCR(uart->swk_index),0x03);	//子串口 发送触发点,接收触发点 
                                                                                    //不使能 发送,接收FIFO 复位发送接收FIFO
        /*子串口 中断使能寄存器*/
        EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE0_SIER(uart->swk_index),0x00); //子串口 禁止接收FIFO数据错误中断
                                                                                    //禁止发送FIFO空中断
                                                                                    //禁止发送FIFO触点中断
                                                                                    //禁止接收FIFO接收超时中断
                                                                                    //禁止接收FIFO接收触点中断
        uart->irq_enable = 0;
        break;
    case RT_DEVICE_CTRL_SET_INT:
        // /* enable rx irq */
        /*切换到PAGE0页中的子串口寄存器组 */
        EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE(uart->swk_index),0x00);
        /*子串口 控制寄存器 */
        EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE0_SCR(uart->swk_index),0x03);	//子串口 发送使能  接收使能
        /*子串口 FIFO控制寄存器*/
        EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE0_FCR(uart->swk_index),0x0F);	//子串口 发送触发点,接收触发点 
                                                                                    //使能 发送,接收FIFO 复位发送接收FIFO
        /*子串口 中断使能寄存器*/
        EXHW_WK2124_Write_Reg(uart->spi_device, SPAGE0_SIER(uart->swk_index),0x83); //子串口 使能接收FIFO数据错误中断
                                                                                    //禁止发送FIFO空中断
                                                                                    //禁止发送FIFO触点中断
                                                                                    //使能接收FIFO接收超时中断
                                                                                    //使能接收FIFO接收触点中断
        uart->irq_enable = 1;
        break;
    }
    return RT_EOK;
}

static int wk2124_putc(struct rt_serial_device *serial, char c)
{
    uint8_t sendbuf[10];
    struct wk2124_uart *uart;
    RT_ASSERT(serial != RT_NULL);
    uart = (struct wk2124_uart *)serial->parent.user_data;

    sendbuf[0] = c;
    Wk2124_SendBuf(uart->spi_device, uart->swk_index, sendbuf, 1);
    while(EXHW_WK2124_Read_Reg(uart->spi_device, SPAGE0_TFCNT(uart->swk_index))  > 0);
    while((EXHW_WK2124_Read_Reg(uart->spi_device, SPAGE0_FSR(uart->swk_index)) & 0x01) == 1);

    return 1;
}

static int wk2124_getc(struct rt_serial_device *serial)
{
    int ch;
    uint8_t recbuf[10];

    struct wk2124_uart *uart;
    RT_ASSERT(serial != RT_NULL);
    uart = (struct wk2124_uart *)serial->parent.user_data;

    ch = -1;    //确保没有接收数据时rt_device_read返回值为0
    if(Wk2124_GetBuf(uart->spi_device, uart->swk_index, recbuf, 1)) {
        ch = recbuf[0];
    }
    return ch;
}

static const struct rt_uart_ops wk2124_uart_ops =
{
    wk2124_configure,
    wk2124_control,
    wk2124_putc,
    wk2124_getc,
};

#if defined(PKG_USING_UART_SWK1)
/* UART1 device driver structure */
static struct wk2124_uart uart_swk1;
struct rt_serial_device serialswk1;
void WK2124_UART1_IRQHandler(void)
{
    struct wk2124_uart *uart;
    uart = &uart_swk1;
    volatile uint8_t uart_irq_stat = 0; 
    RT_ASSERT(uart != RT_NULL);
    RT_ASSERT(&serialswk1 != RT_NULL);
    RT_ASSERT(uart->spi_device->parent.type == RT_Device_Class_SPIDevice);
    
    if (uart->irq_enable == 0) {
       return;
    }
    /*判断串口的中断类型*/
    uart_irq_stat = EXHW_WK2124_Read_Reg(uart->spi_device, SPAGE0_SIFR(uart->swk_index));
    //子串口接收FIFO触点中断标志 ， 子串口接收FIFO超时中断标志
    if(uart_irq_stat & (3 << 0)) {    
        rt_hw_serial_isr(&serialswk1, RT_SERIAL_EVENT_RX_IND);
    }
}
#endif /* PKG_USING_UART_SWK1 */

#if defined(PKG_USING_UART_SWK2)
/* UART2 device driver structure */
static struct wk2124_uart uart_swk2; 
struct rt_serial_device serialswk2;
void WK2124_UART2_IRQHandler(void)
{
    struct wk2124_uart *uart;
    uart = &uart_swk2;
    volatile uint8_t uart_irq_stat = 0; 

    RT_ASSERT(uart != RT_NULL);
    RT_ASSERT(&serialswk2 != RT_NULL);
    RT_ASSERT(uart->spi_device->parent.type == RT_Device_Class_SPIDevice);
    
    if (uart->irq_enable == 0) {
       return;
    }
    /*判断串口的中断类型*/
    uart_irq_stat = EXHW_WK2124_Read_Reg(uart->spi_device, SPAGE0_SIFR(uart->swk_index));
    //子串口接收FIFO触点中断标志 ， 子串口接收FIFO超时中断标志
    if(uart_irq_stat & (3 << 0)) {    
        rt_hw_serial_isr(&serialswk2, RT_SERIAL_EVENT_RX_IND);
    }
}
#endif /* PKG_USING_UART_SWK2 */

#if defined(PKG_USING_UART_SWK3)
/* UART3 device driver structure */
static struct wk2124_uart uart_swk3;
struct rt_serial_device serialswk3;
void WK2124_UART3_IRQHandler(void)
{
    struct wk2124_uart *uart;
    uart = &uart_swk3;
    volatile uint8_t uart_irq_stat = 0; 

    RT_ASSERT(uart != RT_NULL);
    RT_ASSERT(&serialswk3 != RT_NULL);
    RT_ASSERT(uart->spi_device->parent.type == RT_Device_Class_SPIDevice);

    if (uart->irq_enable == 0) {
       return;
    }
    /*判断串口的中断类型*/
    uart_irq_stat = EXHW_WK2124_Read_Reg(uart->spi_device, SPAGE0_SIFR(uart->swk_index));
    //子串口接收FIFO触点中断标志 ， 子串口接收FIFO超时中断标志
    if(uart_irq_stat & (3 << 0)) {    
        rt_hw_serial_isr(&serialswk3, RT_SERIAL_EVENT_RX_IND);
    }
}
#endif /* PKG_USING_UART_SWK3 */

#if defined(PKG_USING_UART_SWK4)
/* UART4 device driver structure */
static struct wk2124_uart uart_swk4;
struct rt_serial_device serialswk4;
void WK2124_UART4_IRQHandler(void)
{
    struct wk2124_uart *uart;
    uart = &uart_swk4;
    volatile uint8_t uart_irq_stat = 0; 

    RT_ASSERT(uart != RT_NULL);
    RT_ASSERT(&serialswk4 != RT_NULL);
    RT_ASSERT(uart->spi_device->parent.type == RT_Device_Class_SPIDevice);

    if (uart->irq_enable == 0) {
       return;
    }
    /*判断串口的中断类型*/
    uart_irq_stat = EXHW_WK2124_Read_Reg(uart->spi_device, SPAGE0_SIFR(uart->swk_index));
    //子串口接收FIFO触点中断标志 ， 子串口接收FIFO超时中断标志
    if(uart_irq_stat & (3 << 0)) {    
        rt_hw_serial_isr(&serialswk4, RT_SERIAL_EVENT_RX_IND);
    }
}
#endif /* PKG_USING_UART_SWK4 */

/* 中断回调函数 */
void WK2124_IRQHandler(void *args)
{
    rt_sem_release(&irq_sem);
}

static void wk2124_irq_thread_entry(void *parameter)
{
    volatile uint8_t g_irq_stat = 0; 

    while (1)
    {
        rt_sem_take(&irq_sem, RT_WAITING_FOREVER);

        /*使能子串口1,2,3,4的时钟*/
        g_irq_stat = EXHW_WK2124_Read_Reg(wk2124_device, GIFR);
        if(g_irq_stat & (1 << 0)){//子串口 1 有中断
    #if defined(PKG_USING_UART_SWK1)
            WK2124_UART1_IRQHandler();
    #endif
        }
        if(g_irq_stat & (1 << 1)){//子串口 2 有中断
    #if defined(PKG_USING_UART_SWK2)
            WK2124_UART2_IRQHandler();
    #endif
        }
        if(g_irq_stat & (1 << 2)){//子串口 3 有中断
    #if defined(PKG_USING_UART_SWK3)
            WK2124_UART3_IRQHandler();
    #endif
        }
        if(g_irq_stat & (1 << 3)){//子串口 4 有中断
    #if defined(PKG_USING_UART_SWK4)
            WK2124_UART4_IRQHandler();
    #endif
        }
    }
}

int WK2124_IRQ_Init(void)
{
    rt_err_t ret = 0;

    /* 设置引脚为输入模式 */
    rt_pin_mode(WK2124_IRQ_PIN, PIN_MODE_INPUT_PULLUP);
    /* 绑定中断，下降沿模式，回调函数名为WK2124_IRQHandler */
    rt_pin_attach_irq(WK2124_IRQ_PIN, PIN_IRQ_MODE_FALLING, WK2124_IRQHandler, RT_NULL);
    
    /* 初始化信号量 */
    rt_sem_init(&irq_sem, "irq_sem", 0, RT_IPC_FLAG_FIFO);
    rt_thread_t thread = rt_thread_create("wk2124_irq", wk2124_irq_thread_entry, RT_NULL, 1024, RT_THREAD_PRIORITY_MAX / 6, 20);
    if (thread != RT_NULL) {
        rt_thread_startup(thread);
    }  else {
        return -RT_ERROR;
    }

    /* 使能中断 */
    ret = rt_pin_irq_enable(WK2124_IRQ_PIN, PIN_IRQ_ENABLE);
    return ret;
}


int hw_wk2124_usart_init(struct rt_spi_device *device)
{
    struct wk2124_uart *uart;
    struct serial_configure config = RT_SERIAL_CONFIG_DEFAULT;
    rt_err_t ret = 0;

#ifdef PKG_USING_UART_SWK1
    uart = &uart_swk1;
    uart->swk_index = 0;
    uart->irq_enable = 0;
    uart->spi_device = device;
    serialswk1.ops    = &wk2124_uart_ops;
    serialswk1.config = config;
    serialswk1.serial_rx  = RT_NULL;
    /* register UART1 device */
    ret = rt_hw_serial_register(&serialswk1, "uartswk1",
                          RT_DEVICE_FLAG_RDWR | RT_DEVICE_FLAG_INT_RX,
                          uart);
    if (ret) {
        LOG_E("rt_hw_serial_register %s error", "uartswk1");
    }
    RT_ASSERT(ret == RT_EOK);
#endif /* PKG_USING_UART_SWK1 */

#ifdef PKG_USING_UART_SWK2
    uart = &uart_swk2;
    uart->swk_index = 1;
    uart->irq_enable = 0;
    uart->spi_device = device;
    serialswk2.ops    = &wk2124_uart_ops;
    serialswk2.config = config;
    serialswk2.serial_rx  = RT_NULL;
    /* register UART2 device */
    ret = rt_hw_serial_register(&serialswk2, "uartswk2",
                          RT_DEVICE_FLAG_RDWR | RT_DEVICE_FLAG_INT_RX,
                          uart);
    if (ret) {
        LOG_E("rt_hw_serial_register %s error", "uartswk2");
    }
    RT_ASSERT(ret == RT_EOK);
#endif /* PKG_USING_UART_SWK2 */

#ifdef PKG_USING_UART_SWK3
    uart = &uart_swk3;
    uart->swk_index = 2;
    uart->irq_enable = 0;
    uart->spi_device = device;
    serialswk3.ops    = &wk2124_uart_ops;
    serialswk3.config = config;
    serialswk3.serial_rx  = RT_NULL;
    /* register UART3 device */
    ret = rt_hw_serial_register(&serialswk3, "uartswk3",
                          RT_DEVICE_FLAG_RDWR | RT_DEVICE_FLAG_INT_RX,
                          uart);
    if (ret) {
        LOG_E("rt_hw_serial_register %s error", "uartswk3");
    }
    RT_ASSERT(ret == RT_EOK);
#endif /* PKG_USING_UART_SWK3 */

#ifdef PKG_USING_UART_SWK4
    uart = &uart_swk4;
    uart->swk_index = 3;
    uart->irq_enable = 0;
    uart->spi_device = device;
    serialswk4.ops    = &wk2124_uart_ops;
    serialswk4.config = config;
    serialswk4.serial_rx  = RT_NULL;
    /* register UART6 device */
    ret = rt_hw_serial_register(&serialswk4, "uartswk4",
                          RT_DEVICE_FLAG_RDWR | RT_DEVICE_FLAG_INT_RX,
                          uart);
    if (ret) {
        LOG_E("rt_hw_serial_register %s error", "uartswk4");
    }
    RT_ASSERT(ret == RT_EOK);
#endif /* PKG_USING_UART_SWK4 */

    return ret;
}


int wk2124_spi_init(const char *spi_dev_name)
{
    RT_ASSERT(spi_dev_name);

    if (wk2124_device != RT_NULL)
    {
        return 0;
    }

    wk2124_device = (struct rt_spi_device *) rt_device_find(spi_dev_name);
    if (wk2124_device == RT_NULL)
    {
        LOG_E("You should attach [%s] into SPI bus firstly.", spi_dev_name);
        return -RT_ENOSYS;
    }

    /* check SPI device type */
    RT_ASSERT(wk2124_device->parent.type == RT_Device_Class_SPIDevice);

    /* configure SPI device*/
    {
        struct rt_spi_configuration cfg;
        cfg.data_width = 8;
        cfg.mode = RT_SPI_MASTER | RT_SPI_MODE_0 | RT_SPI_MSB;  /* SPI Compatible Modes 0 */
        cfg.max_hz = 40 * 1000 * 1000;                          /* SPI Interface with Clock Speeds Up to 40 MHz */
        if (rt_spi_configure(wk2124_device, &cfg)) {
            LOG_E("rt_spi_configure SPI device %s error.", spi_dev_name);
            return -RT_ERROR;
        }
    }

    if (rt_device_open((rt_device_t) wk2124_device, RT_DEVICE_OFLAG_RDWR) != RT_EOK)
    {
        LOG_E("open WK2124 SPI device %s error.", spi_dev_name);
        return -RT_ERROR;
    }

    return RT_EOK;
}



int wk2124_device_init(void)
{
    rt_err_t ret = 0;

#ifdef WK2124_SPI_DEVICE
    rt_thread_mdelay(100);

    ret = wk2124_spi_init(WK2124_SPI_DEVICE);
    if (ret == RT_EOK) {
        rt_thread_mdelay(100);
        WK2124_IRQ_Init();
        EXHW_WK2124_Init(wk2124_device);
        rt_thread_mdelay(100);
        ret = hw_wk2124_usart_init(wk2124_device);
        if (ret != RT_EOK) {
            LOG_E("RT-Thread WK2124 package initialize fail.");
            return -RT_ERROR;
        }
    }  else {
        LOG_E("RT-Thread WK2124 package initialize fail.");
        return -RT_ERROR;
    }
#else
    return -RT_ERROR;
#endif
    LOG_I("RT-Thread WK2124 package initialize success.");
    return RT_EOK;
}

INIT_ENV_EXPORT(wk2124_device_init);

#endif /* PKG_USING_WK2124 */
